// coordinator-node-client/lib/DefaultConfig.js
/**
 * Copyright (c) 2010-2015 EEFUNG Software Co.Ltd. All rights reserved.
 * 版权所有 (c) 2010-2015 湖南蚁坊软件有限公司。保留所有权利。
 * Created by liwenjun on 2016/4/13.
 */

exports.config = {
    instance: {
        "hostName": "localhost",
        "app": "myapp",
        "ipAddr": "127.0.0.1",
        "status": "UP",
        "port": {
            "$": 8080,
            "@enabled": "true"
        },
        "dataCenterInfo": {
            //不要改
            "@class": "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo",
            //不要改
            "name": "MyOwn"
        },
        "vipAddress": "coordinator.nodejs.client"
    },
    coordinator: {
        /*
         coordinator server  domain|| (host&&port)
         domain优先，如果设置了domain就是用domain访问coordinator；
         如果未设置domain或设置为"",则需要设置ip和端口，并使用ip和端口访问coordinator
         */
        domain: "",//例如mydomain.com
        host: "172.19.103.51",
        port: 80
    },
    registerSelf: false,
    agent: {
        open: false,//是否使用代理
        agentOptions: {
            socksHost: '192.168.2.171', // Defaults to 'localhost'.
            socksPort: 10010 // Defaults to 1080.
        }
    }
};
