// caster-client-sdk/util/custom-logger.js
/*!
 * Copyright (c) 2010-2020 EEFUNG Software Co.Ltd. All rights reserved.
 * 版权所有 (c) 2010-2020 湖南蚁坊软件股份有限公司。保留所有权利。
 */

"use strict";

/**
 * log4js 在 caster-client 中的日志代理
 * Created by wuyaoqian on 2016/12/15.
 */

var path = require('path');
var defaultLogger = require("log4js-config").get("eefung.webapp.caster-client." + path.basename(__filename));

// copy from lib/logging/LoggingService
var LogLevel = {};
LogLevel[LogLevel["error"] = 0] = "error";
LogLevel[LogLevel["warn"] = 1] = "warn";
LogLevel[LogLevel["info"] = 2] = "info";
LogLevel[LogLevel["debug"] = 3] = "debug";
LogLevel[LogLevel["trace"] = 4] = "trace";

var CustomLogger = function (logger) {
    this.logger = logger || defaultLogger;
};
CustomLogger.prototype.log = function (level, className, message, furtherInfo) {
    var type = LogLevel[level];
    if (type && typeof type === 'string') {
        if (furtherInfo) {
            this.logger[type]('%s: %s -> Detail:\n', className, message, (furtherInfo.stack ? furtherInfo.stack : furtherInfo));
        } else if (typeof message === 'object') {
            this.logger[type]('%s -> Detail:\n', className, (message && message.stack ? message.stack : message));
        } else {
            this.logger[type]('%s: %s', className, message);
        }
    } else {
        console.debug('CasterClientLogNotFound, level:%s,className:%s,message:%s,furtherInfo:', level, className, message, (furtherInfo.stack ? furtherInfo.stack : furtherInfo));
    }
};

module.exports = CustomLogger;
