// @component-util/micro-pubsub-service/lib/broadcast-service.js
/*!
 * Copyright (c) 2010-2020 EEFUNG Software Co.Ltd. All rights reserved.
 * 版权所有 (c) 2010-2020 湖南蚁坊软件股份有限公司。保留所有权利。
 */

'use strict';

const EventEmitter = require('events');
const {getInstance, constructorCheck} = require('../util/service-util.js');
const BaseService = require('../util/base-service.js');

/**
 * broadcast-service.js
 * Created by wuyaoqian on 2019/1/22.
 */

class BroadcastService extends BaseService {

    constructor (name) {
        super();
        constructorCheck.call(this, new.target, BroadcastService, name, () => {
            this.name = name;
            this.emitter = new EventEmitter();
        });
    }

    /**
     * 广播数据
     * @param {*} data
     * @return {BroadcastService}
     */
    broadcast (...data) {
        if (this.emitter) {
            this.emitter.emit(this.name, ...data);
        }
        return this;
    }

    /**
     * 销毁广播服务
     */
    destroy () {
        if (this.emitter) {
            this.emitter.removeAllListeners(this.name);
            delete this.emitter;
            delete this.constructor.services[this.name];
        }
    }

    /**
     * 发布一个广播服务
     * @param {string} name
     * @return {BroadcastService}
     */
    static publish (name) {
        return getInstance(this, BroadcastService, 'publish', [name]);
    }

    /**
     * 监听广播服务
     * @param {string} name
     * @param {function} fn
     * @return {function}
     */
    static on (name, fn) {
        if (!fn || typeof fn !== 'function') {
            return () => {};
        }
        let bind_fn = fn.bind(null);
        let instance = getInstance(this, BroadcastService, 'on', [name]);
        instance.emitter.on(name, bind_fn);
        return () => {
            instance.emitter.removeListener(name, bind_fn);
        };
    }

}

/**
 * 静态方法：
 * 0. getService             // function(name) { return SubBroadcastService; }
 * 1. publish                // function(name) { return instance; }
 * 2. on                     // function(name, fn) { return selfOffHandler; }
 *
 * 实例方法：
 * 1. broadcast              // function(...data) { return instance; }
 * 2. destroy                // function() { return undefined; }
 *
 * @type {BroadcastService}
 */
module.exports = BroadcastService;
