// @component-util/async-load-js/lib/async-load-js.js
/*!
 * Copyright (c) 2010-2020 EEFUNG Software Co.Ltd. All rights reserved.
 * 版权所有 (c) 2010-2020 湖南蚁坊软件股份有限公司。保留所有权利。
 */

'use strict';

/**
 * async-load-js.js
 * Created by wuyaoqian on 2019/6/14.
 */

const asyncScript = require('scriptjs');

// 设置请求url的host部分值，给直接使用根目录访问之用（不能使用 path， 因为合并后会自动的在ur之后加 .js 进行结尾）
// asyncScript.path(`${location.protocol}//${location.host}`);

const AsyncProcess = {
    running: false,
    queue: [],
    addTask: function (task) {
        AsyncProcess.queue.push(task);
        if (!AsyncProcess.running) {
            AsyncProcess.running = true;
            AsyncProcess.process();
        }
    },
    process: function () {
        let data = AsyncProcess.queue.shift();
        // 执行顺序：
        // 1. set args （ sync ）
        // 2. async download js ( async-1 )
        // 3. clear args ( async-2 )
        // 4. repeat next process ( async-2 )
        if (data) {
            // 1. (sync) 设置请求参数（将参数与路径分隔开的原因是：下次再异步下载同一路径但参数不同时，将不会触发远程异步下载）
            asyncScript.urlArgs(data.args);
            // 2. (async-1) 异步下载（因 asyncScript 内部原理是在 setTimeout 后，才去请求下载）
            asyncScript(data.path, data.key, data.done);
            // 3. (async-2) 异步执行 4, 5
            setTimeout(function () {
                // 4. (sync) 因 asyncScript 是全局唯一，所以 urlArgs 用完后，需要清理
                asyncScript.urlArgs('');
                // 5. (sync) 开启下一个异步下载处理
                AsyncProcess.process();
            }, 0);
        } else {
            AsyncProcess.running = false;
        }
    }
};

/**
 * 异步加载 js 资源
 * @param {string|object|array} url
 * @param {string|undefined} [idOrDone]
 * @param {function|undefined} [optDone]
 */
module.exports = function (url, idOrDone, optDone) {
    let path = url;
    let args = '';
    // 支持 string, object, array 三种格式（其中 object, array 支持 path 与 args 的分离）
    if (url.path) {
        path = url.path;
        args = url.args || '';
    } else if (Array.isArray(url)) {
        path = url[0];
        args = url[1] || '';
    }
    AsyncProcess.addTask({
        path: path,
        args: args,
        key: idOrDone,
        done: optDone
    });
};

// 将 asyncScript 暴露出去
module.exports.$script = asyncScript;
